plugins/snapshots/blockfile/blockfile.go

```diff
@@ -472,6 +492,76 @@ func copyFileWithSync(target, source string) error {
 	defer tgt.Close()
 	defer tgt.Sync()
 
-	_, err = io.Copy(tgt, src)
+	if copySparse {
+		// Let's take the following approach here:
+		// * Create the target file
+		// * Truncate the target file
+		// * Copy the data from the source to the target in 1MB chunks
+		info, err := src.Stat()
+		if err != nil {
+			return fmt.Errorf("failed to stat the source %s: %w", source, err)
+		}
+
+		srcSize := info.Size()
+		if err = tgt.Truncate(srcSize); err != nil {
+			return fmt.Errorf("failed to truncate the target %s: %w", target, err)
+		}
+
+		srcFd := int(src.Fd())
+		tgtFd := int(tgt.Fd())
+		bufferSize := int64(1 * 1024 * 1024)
+		buffer := make([]byte, bufferSize)
+		offset := int64(0)
+
+		for offset < srcSize {
+			dataStart, err := syscall.Seek(srcFd, offset, SEEK_DATA)
+			// whence is SEEK_DATA or SEEK_HOLE, and offset is
+			// beyond the end of the file, or whence is SEEK_DATA
+			// and offset is within a hole at the end of the file
+			if err == syscall.ENXIO {
+				break
+			}
+			if err != nil {
+				return fmt.Errorf("SEEK_DATA on %s failed at %d: %w", src, offset, err)
+			}
+
+			holeStart, err := syscall.Seek(srcFd, dataStart, SEEK_HOLE)
+			if err != nil || holeStart > srcSize {
+				// let's ignore any possible error here and
+				// just copy the whole file
+				holeStart = srcSize
+			}
+
+			pos := dataStart
+			for pos < holeStart {
+				n := bufferSize
+				if bufferSize > (holeStart - pos) {
+					n = holeStart - pos
+				}
+
+				i, err := syscall.Pread(srcFd, buffer[:n], pos)
```

Comment:
use [copy_file_range](https://man7.org/linux/man-pages/man2/copy_file_range.2.html) instead?

Reply:
The problem that I have is that it'd make it linux specific (even more).

Reply:
for macos, I think you could just use [clonefile](https://manpagez.com/man/2/clonefile/)
even for linux filesystems, if reflink is surely supported, ioctl(FICLONE) can be used too..